Unify tag namespace and conventions between Neoforge and Fabric for 1.20.5 MC

TelepathicGrunt opened the pull request:
# Background: 

The modding community has been split between two tag namespaces. The `c` tag and the `forge` tag with one using folders and the other not. Now that Neoforge is a thing and willing to have discussions about switching namespaces for their tags, we are at a point were we can finally unify both loaders to the same namespace for tags and conventions!

Fabric PR: https://github.com/FabricMC/fabric/pull/3310

# Preface:

This large tag PR is not perfect but it is a great first steps toward a unified tag namespace and convention to ease the burden between both loader's ecosystems and allow for datapacks or other ecosystems to hop in as well to our namespace and conventions if they desire.
  
There is a ton of moving parts in this PR. Each of these parts I felt are interconnected and are best done all together in a single to help make this switch to the new namespace/conventions as best as possible. The sections below will help break down the PR and go into detail of each part and the reasons for each part. 
  
If after reading the reasons, you still have concerns or suggestions, please voice them on this PR so we can have a discussion! 

# Goals:

1. Unify both modloaders to the same namespace for tags which is ~~`common`~~ in this PR. **EDIT:** we are switching to using `c` namespace instead.

</br>

2. Add or rename tags to better match each other on both loaders and to match the convention of using folders for better organization. 

</br>

3. All past tags have a modern equivalent in this PR with backwards compat using optional tag entries. ~~No tags were entirely removed.~~ (Quartz and amethyst were removed from storage blocks tag with no replacement because they are not storage blocks)

</br>

4. Datagen translations for modloader common tags under the key format of `tag.<registrypath>.c.<tagname>` where slashes are turned into periods. Modded registries should specify their namespace like so so it doesn't conflict with vanilla registry: `tag.<registrynamespace>.<registrypath>.c.<tagname>`

</br>

5. Log short warning by default to console if old tag namespace is detected in a modder's dev environment. Config exists to either turn it off or make it verbose to say all found legacy tags. (Short dev message on by default)

</br>

6. Log short warning by default to console if untranslated modded item tags is detected in a modder's dev environment. Have config to either turn it off or make it verbose to say all found untranslated modded item tags. (off by default)

# Details And Reasons Of Each Goal:

### Namespace choice:

Edit: For better chance of acceptance, we are going to go with `c` namespace. The `c` namespace pollution is largely unique to Fabric ecosystem so it will be on them to work out how to clean it up on their side. On Neo's side, the `c` namespace is largely a blank slate so we should be ok.

~~There are a few reasons to go with `common` namespace as oppose to the currently established `c` namespace on Fabric.~~
  
</br>

~~First, the word `common` is much more clear to new modders or to players/modpack makers that this namespace is shared across many mods. `c` doesn't actually mean anything to people not previously familiar with it. Which can lead to confusion on what `c` is for. `common` appears best for conveying the meaning of the namespace to people. Not a powerful reason for this but it is a reason.~~

~~Second, the `c` namespace is in disaster in conventions at the moment. I'm not talking within Fabric API. I mean as Fabric ecosystem as a whole. To help show what I mean, here is a sheet of all tag usages from the top 1000 Fabric/Quilt mods on Modrinth as of August 4th, 2023: https://docs.google.com/spreadsheets/d/1KBn2hKOz3nnIYh3KAeqv5vezMwFSDO-tL8TUAKYTiBE/edit?usp=sharing~~

~~Third, the reason we are not sticking with `forge` namespace is because Neoforge is not Forge. And we should be using a modloader-agnostic tag namespace anyway so that if Neforge goes under for whatever reason, the new modloader can pick up the new namespace and not have to fight to change it. Makes it much easier on everyone. Keeping `forge` namespace is not an option.~~

~~In there, you can find some people are using folders under `c`. Some people are not. Some people are properly pluralizing their tag names. Some are not. Basically, it's a mess and I feel this was partially because some modders don't know where to find existing tags and some were not able to figure out what the conventions are when looking at Fabric API's tags. To help facilitate Goal 5, switching off of `c` to `common` makes that goal of notifying people of the new convention we are switching to a million times easier. More info about this logging is in the section below about the logging goal. But yes, this was a big reason for my choice of picking `common`. If we go back to `c` instead, please provide me an alternative solution for notifying people of the new conventions.~~

</br>

~~**NOTE:** I know people are going to point out that `common` is a valid modid unlike `c` and that this makes `common` bad. The truth is, that's actually a non-issue. Say if some random mod is made using `common` modid, the modding community will just ignore that mod anyway and it doesn't impact modpacks or anything. Modders will just tell the owner of that mod to change the modid and that'll be that. If it is a heavy concern still, modloaders themselves can just include a dummy `common` modid mod to reserve that namespace to prevent anyone else from using it. Super simple solution. I will leave the choice of reserving the `common` modid up to the modloaders themselves where they can make a PR after this one if we stick with `common` namespace.~~

</br>

### Renaming and moving tags to same conventions

Fabric's `c` convention was to never use folders. Forge's convention was to use folders. And Minecraft started out with not using folders but then now does with its block mineable tags. 

What it looks like to me is that Minecraft starts with no folders originally but as they grew, they started seeing the value of grouping similar tags together by folders. But they still keep the non-folder tags for backwards compatibility. Forge right away saw that trying to mimic vanilla would be an issue for scalability in a modded ecosystem so they went with folders. Fabric however, stuck with the legacy non-folder convention that vanilla did and this lead to a lot of messiness that lead to the current situation with the `c` namespace tag conventions: https://docs.google.com/spreadsheets/d/1KBn2hKOz3nnIYh3KAeqv5vezMwFSDO-tL8TUAKYTiBE/edit#gid=0

If you check what fabric currently does with dyes, all of them are in the format of `c:<color>_dyes` which just bloats the top level view of the `c` tags: https://github.com/FabricMC/fabric/tree/93d8cb82e85c3d5744716430516daef393cf5815/fabric-convention-tags-v1/src/generated/resources/data/c/tags/items

Instead, if we go with the folder convention, these tags would've be grouped under `c:dyes/<color>` which is a lot more organized and cleaner. It is also far more scalable for a modded ecosystem as more tags can be added to the folder without making the top level view harder to find certain tags. Example of what a folder dye tag looks like:
https://github.com/neoforged/NeoForge/tree/1.20.x/src/generated/resources/data/forge/tags/items/dyes

That is why the convention of this PR is plural + folders. Because of this, many tags from both `c` and `forge` were shuffled around so that tags that mean the same thing on both loaders now matches each other in the `c` namespace so multi-loader modders only need to pull from 1 tag instead of 2.

</br>

### Backwards compat with old tags

I put a lot of work into making sure that **all** old tags from the modloader are added as an optional tag entry in the modern equivalent of the tag. This will allow built jars and datapacks using the old tag names to continue to have compatibility throughout 1.21 Minecraft. Then these backwards compatibility can be deleted in 1.22 Minecraft to clean up this remaining tech debt as everyone had an entire major Minecraft version to update their tags. 

However, because of all the tag renaming and moving around, the tag fields in code were also rename to match the changes. **This means this PR is indeed a breaking change for code references to these fields**. This is why this PR is marked for 1.21 Minecraft as that is a good time for doing breaking changes. The reason I did not keep the old field names nor make dummy fields that references the new fields is this would create a lot of tech debt that I felt the modloaders don't want to take on. If this is a deal breaker and that the dummy fields should be added to keep code compat, let me know and I'll update this PR to not be breaking for code even with the tech debt.

As you read this PR's code, you'll notice I put the optional tag entries at the bottom of all the datagen classes for tags. This is so that the optional entry shows up last in the tag's json (cleaner to me and shows it is not super important to viewers) and also makes it very easy to delete in 1.22 Minecraft. Just select the code block at bottom and delete. Done. No more backwards compat. Hopefully this helps!

More info on individual tags can be found in the Misc section.

</br>

### Datagenning tag translations

This might seem strange at first but it turns out that recipe viewer mods are making use of translations for item tags. I asked about this and another person said their own mod would also benefit from item tag translations. So there is a demand for translations of these tags and currently, every mod that does will need to add translations for all the `c` and `forge` tag entries from the modloaders. Doesn't seem right.

So instead, I added translations for the tags and I figured, since I am here making item tag translations, might as well add translations for the rest of the common tags for completeness. I did not add translations for `minecraft` item tags but if desired, I can update this PR to include datagen translations for Minecraft's item tags. Not sure if this is something a Modloader wants to handle from Minecraft. But at the very least, modloaders should be translating their own tags to help out modders and reduce duplicate translations between mods.

Also, `LanguageProvider.add(<TagKey>, <name>)` is added as a helper method that modders can use to make it easier to datagen tag translations. Just takes the TagKey and translated name and will make the lang entry for you!

</br>

### Log warning for legacy namespace presence

This part is critical to getting modders to update to the new namespace and conventions. Basically, there's no way we can reach out to every modder through Discord or through changelogs as people just update their modloader/minecraft version and start porting. There are too many ways that people can slip through the cracks.

So to have a solution that gets this new namespace and convention change out to all modders that actually need this info, this PR include code that will grab all the tags from all Minecraft registries and check for the presence of the legacy namespace. If it is found, it will log a short message that states:

    Dev warning - Legacy Tags detected. Please migrate your 'forge' namespace tags to 'common' namespace! See net.minecraftforge.common.Tags.java for all tags.
    NOTE: Many tags have been moved around or renamed. Some new ones were added so please review the new tags. And make sure you follow tag conventions for new tags!
    You can disable this message in Neoforge's common config by setting logLegacyTagWarnings to "SILENCED" or see individual tags with "DEV_VERBOSE".

If set to verbose, the above message will print out this format:

    Dev warning - Legacy Tags detected. Please migrate your 'forge' namespace tags to 'common' namespace! See net.minecraftforge.common.Tags.java for all tags.
    NOTE: Many tags have been moved around or renamed. Some new ones were added so please review the new tags. And make sure you follow tag conventions for new tags!
    You can disable this message in Neoforge's common config by setting logLegacyTagWarnings to "SILENCED" or see individual tags with "DEV_VERBOSE".

    Legacy tags:
        TagKey[minecraft:item / forge:pies]
        TagKey[minecraft:item / forge:potatoes]
        TagKey[minecraft:entity_type / forge:trains]

This is set to print only in `DEV_SHORT` mode by default so that mod's development environment after world load will see only the shortened message. The modes are `SILENCED`, `DEV_SHORT`, `DEV_VERBOSE`, `PROD_SHORT`, `PROD_VERBOSE`. Prod is included as an opt-in option in case modloader maintainers want to see how many legacy tags are in use for 1.21 modpacks. The prod options can be removed from PR if desired.

Hopefully this isn't too intrusive. But at least the message says how to turn it off as well as where to find the new tags. This should greatly help speed up modders switching to the new tags and help them see what the desired conventions are. ~~The end goal is a `common` namespace that is not a wild-west of conventions like the `c` namespace is. Will this be successful? I don't know but I have hope that it will. Only way to know for sure is to try this out.~~

</br>

### Log warning for untranslated item tags presence

**NOTE:** This part is a bit more controversial but if desired, this logging can be chopped out of this PR. 

**EDIT:** This is now set to off by default and has to be opt-in by the modder in the config file.

The idea is since recipe viewers and other mods rely on item tags to be translated, a short message will be logged if any are found to be untranslated from mods. I did not do the other tag types because there is no demand for those to be translated and thus, doesn't make sense to push for translated tags of other stuff. Only item tags. 

The code in this PR will grab all the item tags and if running on client side, check if the tag is translated. If it is not, it will log a short message that states:

    Dev warning - Untranslated Item Tags detected. Please translate your item tags so other mods such as recipe viewers can properly display your tag's name.
    The format desired is tag.item.<namespace>.<path> for the translation key with slashes in path turned into periods.
    You can disable this message in Neoforge's common config by setting logUntranslatedItemTagWarnings to "SILENCED" or see individual tags with "DEV_VERBOSE".

Setting it to verbose basically prints out all the item tags like so:

    Dev warning - Untranslated Item Tags detected. Please translate your item tags so other mods such as recipe viewers can properly display your tag's name.
    The format desired is tag.item.<namespace>.<path> for the translation key with slashes in path turned into periods.
    You can disable this message in Neoforge's common config by setting logUntranslatedItemTagWarnings to "SILENCED" or see individual tags with "DEV_VERBOSE".

    Untranslated item tags:
      modded1:pies
      modded2:potatoes
      modded2:paintings/1x1

~~This is set to print only in `DEV_SHORT` mode by default so that mod's development environment after world load will see only the shortened message.~~ The default is `SILENCED` mode and has to be opt-in. The modes are `SILENCED`, `DEV_SHORT`, `DEV_VERBOSE`, `PROD_SHORT`, `PROD_VERBOSE`. Prod is included as an opt-in option in case modloader maintainers want to see how many untranslated modded item tags are in use for 1.21 modpacks. The prod options can be removed from PR if desired.

</br>

### Misc

So for the individual tags in this PR, there's a lot of changes done. This is not a comprehensive list but some notable points.

- Added Cherry Grove to biome tags. Yeah it was missing lol. I also added a `c:is_floral` biome tag to hold biomes that are super dense in flowers as quite a few mods are trying to add their stuff to flowery based biomes.

</br>

- Added "vegetation" to the is_sparse and is_dense biome tag names so it is now `is_vegetation_sparse` and `is_vegetation_dense`. Much more clear that the sparse and dense is talking about the density of plants. Not density of other stuff like boulders lol. Makes it easier to match up with the Fabric PR as well.

</br>

- Added `c:honey` fluid tag as honey is added by several mods and is in dire need of standardizing. 

</br>

- `glass` is now `glass_blocks` to be clearer that it is about the full block form of glass. Not a tag of glass blocks and glass panes together.

</br>

- Added `c:boats` and `c:minecart` tags because we probably should have these tags for better compat lol. Surprised they weren't made before.

</br>

- Moved `forge:shears` to `c:tools/shears` because Shears are a tool. Not sure why it was classified as not before.

</br>

- `c:relocation_not_supported` block ~~and block entity type tags are~~ is added because there's many mods that were trying to add their own `forge:relocation_not_supported` or `forge:immovable` tags before. Now it is standardized and should help with adoption to ease mod compatibility.

</br>

- `forge:glass/silica` is renamed to `c:glass/cheap` because my understanding of the silica tag was it was really for checking if a glass block was cheap to be made to help prevent consumption of more expensive glass blocks for various mechanics. So changing the term from silica to cheap should make this far more clear and hopefully better uptake. Especially as a mod could have a glass block that is made from say Sand + Nether Stars to create Star Glass. They would've put this expensive glass into the silica tag because it used Sand but then defeats the original purpose of this tag. Hence why now it is using `cheap`.

</br>

- Brand new `c:dyed_blocks` and `c:dyed_items` tag was added and has a folder for individual colors in it. I added this because the Forge tags had 32 colors tags for Stained Glass and Stained Glass Panes. It's awful. Imagine if this was extended to every colored block... My solution was to remove those colored glass tags and instead, create new overarching tags that collect all dyed color blocks and items. And the subfolder for these dyed_blocks and dyed_items lists each individual color. The goal is now if you wanted Red Stained Glass Blocks, you would check if a block is in the `c:glass_blocks` tag and is also in the `c:dyed_blocks/red` tag. If so, there's your collection of Red Stain Glass Blocks! This is now universal and can be used for any dyed color block! Much cleaner and far more useful. I hope color-based mods like Spectrum will make good use of this!

</br>

- `c:capturing_not_supported` entity type tag now added because there's over 20 mods that I know of that lets players carry an entity or stuff them into an item to move around. I added one entity myself and had to add it to so many mod's disallow tags for picking up into items. We definitely need to standardize this so entity creators only have 1 tag to add to and other mods can pull this tag into their own disallow tag for their stuff.

</br>

- Not every tag from the Fabric PR is ported over to this PR. I only ported the tags that I assumed would be very useful to have such as `c:boats` and `c:minecarts`. The reason for this is some tags in the other loader could be questionable so rather than bog this PR down with arguments over these various tags, I'll leave it up to another person's PR to attempt to add tags from one loader to another after my PR. All I did was make sure that the tags in each loader follows a convention and tags for the same thing matches up.

</br>

# Conclusion

Again, please let me know your feedback or suggestions or thoughts about the various parts of this PR. It is not perfect and getting loaders to use the same namespace and convention for tags is a controversial topic. That is why I am making this PR early long before 1.21 so we can have enough time for discussions.

Please also checkout the discussions on the PR for the other loader too as what is talked about there could also impact this PR. No matter what namespace is chosen in the end, both PRs will use the same namespace. That is a hard requirement of this PR so please make sure others from the other loader are involved in the discussion.

Fabric PR: https://github.com/FabricMC/fabric/pull/3310

Thank you for you time to read this! I know it was long. But this is very important for the modding ecosystems as what is decided for these PRs will have long-lasting impact.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/neoforged/NeoForge?pullRequest=135) <br/>All committers have signed the CLA.

TelepathicGrunt commented:
Note with my latest push I just did:  `common:capturing_not_supported` entity type tag now added because there's over 20 mods that I know of that lets players carry an entity or stuff them into an item to move around. I added one entity myself and had to add it to so many mod's disallow tags for picking up into items. We definitely need to standardize this so entity creators only have 1 tag to add to and other mods can pull this tag into their own disallow tag for their stuff.

Picture of my pain with this lmao: https://imgur.com/svORGdt

Commoble commented:
I'm not translating my tag keys, I don't want logspam in dev for untranslated tag keys, and I don't want to have to change the client config each time I set up a new workspace to not log untranslated tag keys

moonfather1 commented:
you are being tame about notifying people, TG. i suggest a proper warning screen, like the one for experimental worldgen (full screen, white letters on dirt background).

it's not too punishing for people to click one button (dev env only of course), but would drastically help with visibility.

Commoble commented:
Given that mojang isn't translating its own tags, I don't think pushing devs toward making translation keys for theirs isn't something that neoforge should be obligated to do

TelepathicGrunt commented:
Updated PR to now have the item tag translation warning be off by default and modders have to opt-in through the config. The legacy tag warning will direct people to the config anyway and there the modder can choose to have their item tags monitored for missing translations.

Reminder, still good practice to translate your item tags so recipe viewers and other modders can benefit from the tag translations! People do want to follow vanilla's example of hiding resourcelocations from player unless player opts-in to F3 or advanced tooltips. So more tag translations leads to more immersive modpacks with regards to names of tags displayed to players in various modded UI.

@moonfather1 I dunno. I'm on the fence on that if it is too obtrusive. If enough people tell me they are in favor of that, my hard condition for it is that it only shows up once ever per workspace and never again once modder clicks through the window. Only shows for legacy tag warning and only in dev environment of course. Also someone will need to help me as I stink with getting UI/screens to work lol

Commoble commented:
The logspam shouldn't be opt-in, it shouldn't be in neoforge at all. If people want to warn themselves about their own tags, they can write their own warning system.

Speiger commented:
@TelepathicGrunt I am not a really big fan of the "translatable tag keys".
I get where you come from but on the other hand there is a easy solution to all this.
Mods simply can apply a pascal case reggex/formatter on to the resourcelocation system and it will look decent enough.
That is what i do for example for dimension Ids to name.
Simply treat: "_" as spaces and apply a first letter up logic and you get a good output out of 90% of the tags.

One thing that could be standardized is:
The option to allow for translations for the tags, but it should be "optional improvement" System and the default should just apply the pascal converter.
If it looks bad people can provide resource packs through CF/Modrinth that covers all the ones they want improved.

Here is the standard implementation that i use for ALL my mods.
```java
	public static MutableComponent dimension(ResourceKey<Level> dimension) {
		String key = Util.makeDescriptionId("dimension", dimension.location());
		Language lang = Language.getInstance();
		return lang.has(key) ? translate(key) : literal(toPascalCase(dimension.location().getPath()));
	}
	
	public static String toPascalCase(String input) {
		StringBuilder builder = new StringBuilder();
		for(String s : input.replaceAll("_", " ").replaceAll("-", " ").split(" ")) {
			builder.append(firstLetterUppercase(s)).append(" ");
		}
		return builder.substring(0, builder.length() - 1);
	}
```
(Note this implementation expects the server knows the key at least in english, at least if you want to send server chat messages. If the logic is client sided executed it won't rely on the server knowing the translation)
(Edit: I know the toPascalCase implementation could be optimized a lot)

Edit2:
Forcing all devs to support something they don't want could lead to malicious compliance.
What stops the dev to just write a script that auto adds "unsupported" or pascal auto generated to all unknown tags by default to prevent this error from even showing up?
(Not saying that that will happen, but this middle ground suggestion has the upside to basically allow devs that want to implement translation support while others that don't want to to simply ignore it without feeling attacked for it)

TelepathicGrunt commented:
@Speiger counterpoint, EMI mod already standardized how the translation keys are done and already warns devs in dev environment about untranslated tags. So following EMI’s example, that’s the way to go. EMI’s standard is already quite common on Fabric so it doesn’t make sense to start a new competing standard out of nowhere that will cause clashes just like we are experiencing with tags. We are not repeating the same mistakes

Second, translations should be specified and not created through regex so people of other languages can actually make translations. Also remember, not everyone represents spaces in tag names with _ in their resourcelocations and some tag names don’t represent the tag itself like forge:rods actually being Sticks. We want to help people with accessibility too as narrator would perform much better with dedicated clean translations than guesswork regex and narrator will flop with resourcelocations itself

third, this isn’t forced on anyone. The dev has to **opt-in** to the neoforge config to even get the log warning about untranslated tags. And if someone malicious gives bad translations to try and mess with recipe viewers and stuff, that modder will just get bug reports or pack makers shipping a resourcepack to fix the translation. 

The current desired standard is show resourcelocations by default and if a modder wishes to, the modder can provide a translation to show instead to help with immersion and narrator accessibility

Speiger commented:
@TelepathicGrunt 
> The current desired standard is show resourcelocations by default and if a modder wishes to, the modder can provide a translation to show instead to help with immersion and narrator accessibility

Your current implementation doesn't actually represent what you state.

Since you only provide a translation key and not a component you insinuate that the translation key always will work and make moddevs responsible in ensuring that the backup system works.

If you want to implement a "optional" standard then at least provide a implementation that also handles the "optional" part of it.


The rest of the answer has nothing to comment to.
It would just loop.

Anyways last answer from my side.

eerussianguy commented:
My one question is say I have a metal that’s not in vanilla but for consistency I include it — eg `common:ingots/tin`. I guess the solution is for every mod to translate it on their own (and it’s not like there’s much argument over the translation) but it’s worth thinking about

Random832 commented:
Is it that important for the "non-pretty" option to be a resource location rather than a raw translation key? I don't see something like "itemtag.common.ingots.tin" as obviously worse than "#common:ingots/tin"

KnightMiner commented on src/main/java/net/minecraftforge/common/Tags.java:
```diff
@@ -36,15 +39,35 @@ private static void init(){}
         public static final TagKey<Block> COBBLESTONE_INFESTED = tag("cobblestone/infested");
         public static final TagKey<Block> COBBLESTONE_MOSSY = tag("cobblestone/mossy");
         public static final TagKey<Block> COBBLESTONE_DEEPSLATE = tag("cobblestone/deepslate");
+        public static final TagKey<Block> DYED_BLOCKS = tag("dyed_blocks");
```
Why `dyed_blocks` instead of just `dyed`? Partity with the item tag seems more important to me than including the registry name in the tag name.

TelepathicGrunt replied:
Only issue I have with removing `_blocks` is the base tag will then become just `dyed.json` which might seem odd as a tag name. Or is that a preferable name actually?

KnightMiner commented on src/main/java/net/minecraftforge/common/Tags.java:
```diff
@@ -177,7 +184,23 @@ private static void init(){}
+         * Blocks should be included in this tag if their movement can cause serious issues such as world corruption
+         * upon being moved, such as chunk loaders or pipes, for mods that move block entities.
+         */
+        public static final TagKey<BlockEntityType<?>> RELOCATION_NOT_SUPPORTED = tag("relocation_not_supported");
```
What is the difference between this and the block tag? Are mods expected to check both before relocating, or am I required to tag both my block and my BE?

I would recommend having the java docs link to each other for discoverability

TelepathicGrunt replied:
I'll add javadoc (good catch there) but some mods might have multiple block entities based on the same block but not every BE is relocatable. And someone could have a non-BE block that isn't relocatable. Yeah this means the checker has to check block tag first. Then see if it is a BE, then check BE tag next for maximum compat with how other mods might had setup their blocks. I did find there is usage right now of both a block and block entity type tag so definitely linking the two for better visibility is the way to go. 
![image](https://github.com/neoforged/NeoForge/assets/40846040/454b29b7-e15d-4792-ae6c-edcaef86061b)
![image](https://github.com/neoforged/NeoForge/assets/40846040/c47637e3-1c96-4a04-a6bb-ccb81a9936a7)

KnightMiner commented on src/main/java/net/minecraftforge/common/Tags.java:
```diff
@@ -186,10 +209,13 @@ public static class EntityTypes
         public static final TagKey<EntityType<?>> BOSSES = tag("bosses");
+        public static final TagKey<EntityType<?>> MINECARTS = tag("minecarts");
+        public static final TagKey<EntityType<?>> BOATS = tag("boats");
+        public static final TagKey<EntityType<?>> CAPTURING_NOT_SUPPORTED = tag("capturing_not_supported");
```
This should have a javadoc similar to the relocation not supported tag explaining what you mean by "capturing"

KnightMiner commented on src/main/java/net/minecraftforge/common/Tags.java:
```diff
@@ -210,12 +238,30 @@ private static void init(){}
         public static final TagKey<Item> COBBLESTONE_INFESTED = tag("cobblestone/infested");
         public static final TagKey<Item> COBBLESTONE_MOSSY = tag("cobblestone/mossy");
         public static final TagKey<Item> COBBLESTONE_DEEPSLATE = tag("cobblestone/deepslate");
+        public static final TagKey<Item> CLUSTERS = tag("clusters");
```
i feel like clusters might be an unintuitive enough name that mentioning "like amethyst" in a javadoc would help. Same with "budding blocks" and "buds"

KnightMiner commented on src/main/java/net/minecraftforge/common/Tags.java:
```diff
@@ -53,48 +76,19 @@ private static void init(){}
-        public static final TagKey<Block> GLASS_PANES_YELLOW = tag("glass_panes/yellow");
+        public static final TagKey<Block> GLASS_PANES_COLORLESS = tag("glass/panes_colorless");
+
+        public static final TagKey<Block> GLASS_STAINED_BLOCKS = tag("glass/stained_blocks");
```
This seems like an unintutive inconsistency. Why are "colorless" listed before "pane"/"block" but "stained" is listed before? Why not do "glass/blocks/stained" and "glass/blocks/colorless"?

Alternatively, why not `glass/stained` and rely on tag intersection to handle blocks vs panes?

TelepathicGrunt replied:
Good idea with tag intersection. That does solve a pain point I had with these dang glass tags. 

We do have the dyed_blocks tag so the stained tags are actually entirely uneeded

KnightMiner commented on src/main/java/net/minecraftforge/common/Tags.java:
```diff
@@ -53,48 +76,19 @@ private static void init(){}
+        public static final TagKey<Block> GLASS_BLOCKS_CHEAP = tag("glass/blocks_cheap");
+        public static final TagKey<Block> GLASS_BLOCKS_TINTED = tag("glass/blocks_tinted");
 
         public static final TagKey<Block> GLASS_PANES = tag("glass_panes");
```
Shouldn't this be `glass/panes` for consistency wtith the other glass tags? Same with `glass_blocks` being `glass/blocks`

Alternatively, doing `glass_blocks/colorless` and `glass_blocks/stained` makes more sense here.

KnightMiner commented on src/main/java/net/minecraftforge/common/Tags.java:
```diff
@@ -36,15 +39,35 @@ private static void init(){}
+        public static final TagKey<Block> DYED_BLOCKS_WHITE = tag("dyed_blocks/white");
+        public static final TagKey<Block> DYED_BLOCKS_YELLOW = tag("dyed_blocks/yellow");
         public static final TagKey<Block> END_STONES = tag("end_stones");
         public static final TagKey<Block> ENDERMAN_PLACE_ON_BLACKLIST = tag("enderman_place_on_blacklist");
```
This tag has functional behavior rather than being for commonality between mods, it should remain in `forge` or `neoforge`.

We should probably also javadoc it, but since its an old tag I would not say its this PR's job to do it.

KnightMiner commented on src/main/java/net/minecraftforge/common/Tags.java:
```diff
@@ -177,7 +184,23 @@ private static void init(){}
 
```
`needs_x_tool` should all be left in `forge`/`neoforge` as they control behavior in forge, they are not for commonality betwen mods.

KnightMiner commented on src/main/java/net/minecraftforge/common/Tags.java:
```diff
@@ -210,12 +238,30 @@ private static void init(){}
         public static final TagKey<Item> CROPS_NETHER_WART = tag("crops/nether_wart");
         public static final TagKey<Item> CROPS_POTATO = tag("crops/potato");
         public static final TagKey<Item> CROPS_WHEAT = tag("crops/wheat");
+        public static final TagKey<Item> DYED_ITEMS = tag("dyed_items");
```
Same as the blocks, I see no need to mention the registry name. `dyed` is sufficient, and provides parity with the block tag name. The name right now is confusing as it might suggest that glass *blocks* do not belong in the tag, when I assume its meant to hold any items that might be considered dyed.

TelepathicGrunt replied:
Hmm. that is a good argument. I'll make the change and we will see if people are cool with the base tag name being just `dyed.json`

KnightMiner commented on src/main/java/net/minecraftforge/common/Tags.java:
```diff
@@ -260,48 +306,19 @@ private static void init(){}
         public static final TagKey<Item> GEMS_PRISMARINE = tag("gems/prismarine");
```
`enchnating_fuels` is another tag that should be left under `forge`/`neoforge`. Its not changed in the diff so Github Won't let me comment on it directly.

KnightMiner commented on src/main/java/net/minecraftforge/common/Tags.java:
```diff
@@ -453,14 +480,23 @@ private static void init(){}
          */
         public static final TagKey<Item> TOOLS_FISHING_RODS = tag("tools/fishing_rods");
```
I think we really should ditch all the tool tags in favor of tool actions as the tags disallow NBT sensitive implementation of tools (which I do constantly). It would be trivial to implement a tool action ingredient, but that idea was turned down without discussion when I brought it up before. While we are breaking tags might as well discuss again.

I know the tags say "this is not an alternative or a substitute to tool actions" but that is an absolutely useless comment without elaboration on how the two are meant to be used differently. In my experience, the two should be 100% equivelent barring limitations of tags.

FiniteReality replied:
I think we should keep them, but definitely encourage functional behaviour to be done using tool actions. Tagging tool types is, in my eyes, more important for crafting recipes and slot interactions (like an automatic fisher which accepts any fishing rod, using its stats to affect how many fish are produced) than actually using the tool.

KnightMiner replied:
You can use a tool action for a crafting recipe or for a slot interaction.

TelepathicGrunt replied:
Fabric has these tool tags as well. I am not seeing a tool action in the vanilla recipe json
![image](https://github.com/neoforged/NeoForge/assets/40846040/4f0f3f9c-f3a9-482f-9854-f6b449ebaeb6)

If it is a modded thing, then not having these tools defined by tag makes it harder for datapackers outside Neoforge to make use of modded tools if their datapack gets used on Neoforge. They could have a chest that uses a tag entry to randomly generate any modded or vanilla fishing rod. I too think we we should keep these tool tags for both parity with fabric and help out datapackers.

~~But, I will definitely add javadoc so that neoforge modders know to use tool actions instead for determining a tool's behavior.~~

Looks like there's already a javadoc. Might adjust wording to be more clear
![image](https://github.com/neoforged/NeoForge/assets/40846040/bc589f96-93ef-439c-a3f3-fb52a043be6a)

KnightMiner replied:
Yes, my original comment said the javadocs is pretty terrible. It just states it's "not the same" but states nothing about the difference

TelepathicGrunt replied:
Ah then yeah this new javadoc should help get neo modders to use the ToolActions rather than the tag for determining behaviors
![image](https://github.com/neoforged/NeoForge/assets/40846040/714b52c9-9b46-4035-9b95-4f17c504a684)

KnightMiner commented on src/main/java/net/minecraftforge/common/Tags.java:
```diff
@@ -453,14 +480,23 @@ private static void init(){}
+         * @see net.minecraftforge.common.ToolAction
+         * @see net.minecraftforge.common.ToolActions
+         */
+        public static final TagKey<Item> TOOLS_SPEARS = tag("tools/spears");
```
Is this for spears or throwable weapons? those are two very different things. Is a boomerang, a throwing knife, or a shuriken a "spear"? I would also call a quarterstaff a spear despite it not being throwable.

TelepathicGrunt replied:
This one is am having trouble with for the name but basically, if I name this tridents, it might prevent people from adding their throwable impaling stick weapon into here. So I broadened the name from trident to spears because tridents is a subset of spears (it's a 3 prong spear by definition)

Boomerangs and other throwing stuff aren't spears and don't belong in this tag as they aren't spears. Instead, modders for those should make their own tool tag for those kinds of weapons as someone querying spears to get a boomerang or throwing knife is going to be very confused.

What I can do I expand the javadoc to explicitly state it is for throwing spears like tridents and anything else is best made into their own tag.

also remember, this is not about all throwable weapons (that's what the tool action is for) but for specifically throwing spears for say jungle fortress random loot immersion.

moonfather1 replied:
for spears, i would suggest `trident_like`. yes i agree that tridents are sub-type of spear, but mc game introduced tridents first so it makes sense to call the tag trident_like. similar modded weapons (spears, bidents) will be based on trident.

TelepathicGrunt replied:
I thought some more and yeah `spears` is totally fine. Example, if someone makes a Sword that is throwable like a Trident, it should still go in the `swords` tag as it is a sword, not a trident despite having the same behavior of a trident. These tags are classifying the kind of tool, not the behavior of these tools. 

Thus even non-throwable spears should be allowed here as well. I went ahead and removed the `throwable` part of the javadoc out so now it should be clear that this is about the tool itself and not the behavior.

(For some background, the tag was originally `c:spears` on Fabric so they have been running with that for awhile and haven't been having any problems)

If that's good with y'all, I'll mark this convo as resolved

KnightMiner commented on src/main/java/net/minecraftforge/common/Tags.java:
```diff
@@ -500,10 +536,14 @@ private static void init() {}
+        /**
+         * Holds all fluids related to honey.
+         */
+        public static final TagKey<Fluid> HONEY = tag("honey");
```
Inclusion of honey here feels weird without also including soups, stews, and alike; all are fluids that forge does not implement but are common in mods. Potion is also a common fluid tag.

Also, since tags are meant for recipe equivalencies, I think any Forge provided fluid tags should really mention units. The standard for `forge:honey` is 250mb per bottle, anyone who does not follow that standard will cause duplication bugs in the mods that use the tag so they should not use the tag. Same would be worth mentioning for soups and potions (250mb per bottle/bowl). Milk probably does not need this as a bucket of milk is clearly defined in vanilla.

TelepathicGrunt replied:
Reason I added honey fluid tags was because Fabric actually has a honey fluid tag and that I know of 7 mods currently in 1.19.2+ that has honey fluids (Create, Inspiration, The Bumblezone, Cyclic, Tinkers, Productive Bees, Resourceful Bees)

Though I don't understand the soup and potion thing. Potion is its own item. And soup items can be holding any kind of fluid (like rabbit stew vs mushroom stew fluids)

Or is it that people are converting these items to fluids like a general `soup` fluid for all stews and a `potion` fluid for all potions and sticking that into a bucket? And how many mods are doing those things?

KnightMiner replied:
Tinkers, Thermal, and a few other mods have a potion fluid, 250mb per bottle. Create also has a potion fluid but they use a different standard.

Mushroom soup, beetroot soup, and alike are all fluids in inspirations as well, and I've seen them in another mod before.

KnightMiner replied:
To clarify, all potions are generally 1 fluid, using NBT. Everyone agreed on a NBT format. They get used either when the mod supports fluid crafting or when the mod wishes to consume fluids for some effect.

Each soup is a separate fluid, because there are not that many soups (just 3) and its not backed by another registry. Its just about parity as a lot of people like treating "vanilla thing that is a fluid in a container" as a fluid.

TelepathicGrunt replied:
Forgot to ask, are bowls treated like bottles of 250mb per bowl?

TelepathicGrunt replied:
I'll go with bowls are 250mb for now. I added the fluid tags you asked for plus suspicious stew. lmk if the bowls are actually a different measurement and I'll fix the javadoc

KnightMiner commented:
Left a bunch of comments for discussion, while I overall agree with the changes there were a few changes to tags that seemed arbitrary without making similar tags, and a few cases where we have used tags despite them being the wrong tool that might be worth addressing while we already are breaking tags.

I left a couple of comments stating I oppose moving tags related to Forge logic to the common namespace, those should be left separate as its not meant for unification, its meant to implement forge behavior. I might have missed some, I recommend just doing a find usages for the forge tags class outside datagen; any tag that is used outside datagen should either be moved to forge or split into a forge tag containing a common tag in my opinion. Also note when I say "forge" here, either `forge` or `neoforge` is fine, whatever the team prefers for forge logic moving forwards.

TelepathicGrunt commented:
Forgot to say, for anyone trying to review this PR through GitHub website but is lagging super hard, click on `file changed` tab, then click `file filter`, and then uncheck `json`. This will hide all the datagenned files and stop GitHub from lagging anymore and let you easily do the review. 
![image](https://github.com/FabricMC/fabric/assets/40846040/62f9a751-79e5-4fb5-a048-b04f5e8cd30c)

KnightMiner commented on src/main/java/net/minecraftforge/common/data/ForgeItemTagsProvider.java:
```diff
@@ -23,11 +23,20 @@
 import java.util.concurrent.CompletableFuture;
 import java.util.function.Consumer;
 
+import static net.minecraftforge.common.Tags.Blocks.*;
```
Why is there both a wildcard static import and individual statc imports?

TelepathicGrunt commented:
> My one question is say I have a metal that’s not in vanilla but for consistency I include it — eg `common:ingots/tin`. I guess the solution is for every mod to translate it on their own (and it’s not like there’s much argument over the translation) but it’s worth thinking about

Yeah each mod would provide their translation for it. If so many people are using the same tag, one of the modders can then try making a PR into Neo to make it a default and Neo provides the tag and translation. How many mods using a non-standard tag before it'll get accepted into mod-loader? That's probably a discussion for another day or PR lol. But lets say 6 mods for now as when 6 are using the same tags, maybe it should be standard?

TelepathicGrunt commented:
> Is it that important for the "non-pretty" option to be a resource location rather than a raw translation key? I don't see something like "itemtag.common.ingots.tin" as obviously worse than "#common:ingots/tin"

Generally recipe viewers will fallback to displaying the resourcelocation as that was the original standard. Basically they want to move from resourcelocation standard to translated standard but fallback to the old way if no tags as users will be expecting either format. 

But users won't be expecting the untranslated tag key being visible as that looks like a bug and can lead to mods getting bug reports. I want to keep the translated tag as something more optional and opt-in rather than influencing players to start making bug reports to mods. So it seems ideal to stick with resourcelocation if no translations as users will not report that to modders. Maybe in future after larger tag translation uptake, we can think about showing the untranslated tag key instead to push harder for translations but for now, lets try and keep it more optional and less likely to have players bugging modders.

TelepathicGrunt commented:
I have pushed up several big changes and fixes for this PR. The biggest change is

## This PR is now using `c` tags.
While this may not be ideal for some people, having these tag PRs get accepted by both communities is very important to getting an accepted single standard across loaders. However, it will be `c` with the folder convention as the basis of this PR. What I seen, communities, even outside of Fabric, are willing to switch to `c` tags as long as the convention keeps folders.

Please review this PR and let me know if any tag names or folder names are incorrect. If there is a mistake anywhere. Etc

TelepathicGrunt commented:
New big question as this resolves around the `storage_blocks` tag folder. It seems like a good idea on the surface but it has many glaring issues I think needs to be addressed but I need multiple people's feedback on which way to go with it.

1. It contains Amethyst and Quartz blocks which are not storage blocks. Once you craft the block, you cannot break it back down to the item resource again. Effectively, it's a resource sink. Not storage.

2. There's no way to tell what the compression rate is of these tags. If we remove amethyst and quartz, all the blocks in the folder have a compression rate of 3x3. but what if someone adds a 2x2 compression tag into it? Or a 4x4? 

3. If we remove amethyst and quartz but people still want to query for the existence of those blocks, what can they do?

Some ideas.

Create a `material/` tag folder where we state basic materials like redstone, quartz, etc ignoring any compression rates. That way those that just want blocks that are made mainly of iron, can detect those blocks for their behaviors. Like have a special lightning that chains through any material iron blocks. It would have both block of iron and raw block of iron in it. Then we can safely remove amethyst and quartz out of `storage_blocks` to make it purely 3x3 compression true storage blocks. 

Alternatively, in addition to the material tag idea, we just get rid of `storage_blocks` entirely and provide an in-api way of determining if a block has a compression recipe and give all the compression rates and resources that the block has to the modder. Then modders can know exactly what the compression rate is and what the resources possible are for any given block if it does have a recipe both ways.

KnightMiner commented:
I use the storage blocks tags extensively for metals, getting rid of it is not a good idea. I doubt I'm the only one adding recipes that take 3x3 blocks of metal.

I commented on a forge PR awhile back about the naming issue and the compression rate issue. Don't recall the names I suggested, but they basically assumed that 3x3 blocks were reversible and 2xw were not, plus clay bricks got included in the 2x2 blocks.

TelepathicGrunt commented:
Yeah I would like to see storage_blocks only hold reversible blocks. Doesn't make sense to have none reversals. Maybe this is a case where subfolders is required? Like `storage_blocks/3x3/iron`  and now the compressibility is known which helps gets rid of some assumptions about the tag.

KnightMiner commented:
Personally I'd rather see just a number instead of a shape, e.g. `storage_blocks/nine` and `storage_blocks/four`. That said, do we really expect any sizes beyond the two?

That said, I cannot think of any reversible block in vanilla that is 2x2, unless you count blocks reversible via their loot table (clay, snow, etc.). And given there are only two sizes common enough to warrant tags, I would argue that `storage_blocks` be listed as always 9 to 1, and tell people to come up with another name for 2x2/suggest one in the javadocs.

For non-reversible 2x2, `compressed_blocks` is possibly an option. Would technically include bricks to brick block, sand to sandstone, quartz, and alike. Snow, clay, and alike are debatable

TelepathicGrunt commented:
@KnightMiner 
> That said, I cannot think of any reversible block in vanilla that is 2x2

Honey blocks technically fits the bill as it is 2x2  and has a reversing recipe.

but the glass bottles in the reverse recipe requires 5 slots. Which just complicates this more lol

KnightMiner commented:
Honey blocks are odd enough that unless we see a second that works like it, I'd just call it an outlier and not handle it specifically in tags.

TelepathicGrunt commented:
Summary of new changes:

Amethyst and Quartz is removed from `storage_blocks/` folder because they are not storage_blocks. If a mod adds a proper reversing recipe for them, the mod can add the storage_blocks tag back. (ideally under a `small_storage_blocks/` folder since they are 2x2). But that'll be up to modders themselves. No `materials/` tag added. Would be too much of a rabbit hole of what to support.

Fluid tags expanded to include potion, stews, and soups.

Enchantment tags ported from Fabric PR to here

TelepathicGrunt commented:
Added Bone, Dried Kelp, Hay, and Slime storage blocks that were missing. Names of these tags can be changed as Bone, Hay, and Slime can be tricky due to item and block name not matching up nicely. Need feedback on this for if storage blocks is based on block's name or resource's name

TelepathicGrunt commented:
`c:relocation_not_supported` Block Entity Type tag is now removed in favor of having people use the block tag instead. Reason for this is I got in contact with 4 modders that were using the Block Entity Tag and of the 3 that got back to me, they all said they are cool with switching to the block tag instead. So to simplify checking the tags, the Block Entity Type tag is removed so movers only have to check 1 tag now.

KnightMiner commented on src/main/java/net/minecraftforge/common/Tags.java:
```diff
@@ -500,10 +584,72 @@ private static void init() {}
+         */
+        public static final TagKey<Fluid> HONEY = tag("honey");
+        /**
+         * Holds all fluids related to potions. The effects of the potion fluid should be read from NBT.
```
I recommend a link to the `PotionUtils` method we all use here, its the one that reads a potion from a tag compound. A general link to potion utils is probably fine alternatively.

KnightMiner commented on src/main/java/net/minecraftforge/common/Tags.java:
```diff
@@ -500,10 +584,72 @@ private static void init() {}
          * Holds all fluids that are gaseous at room temperature.
```
The one other fluid I wonder if its worth having is powdered snow, though I suspect that one is pretty uncommon (I have not even gotten around to including it in my mods). The fact it has a bucket form might be enough for some mod devs though.

TelepathicGrunt replied:
I would say powder snow is not a fluid as otherwise, sand can be considered a fluid if someone make a bucket for it which doesn't really make sense and that powder snow bucket places a block instead of fluid.

If I was doing a tech mod with fluid stuff, I would design my system to make people pump water around instead of powder snow. Then have a freezer block that takes in water and buckets and freezes the water into powder snow buckets and spit that out. That would be more logical imo.

So I'll leave out any sort of tag for powder snow fluid. Let mods decide if they want to add that tag on their own or not

TelepathicGrunt replied:
Gonna mark this as resolved for now but if we start seeing lots of mods with powder snow fluid for some odd reason, we can revisit this.

TelepathicGrunt commented:
Just an update, someone ask about a meat and fish tag for foods which sparked some discussion on both loaders. So I ran a poll on what the name show be for the fish and meat tags. Ultimately, the plural names of `meat` and `fish` won instead of `meats` and `fishes`.
https://strawpoll.com/jVyG8KQYYn7

Then someone pointed out it should probably be under the food folder. So I ran a second poll for if it should be `foods` or `food`. The `food` folder and tag name won out.
https://strawpoll.com/NoZr3lLrGy3

Hence why the recent additions of these food tags are named like such. The polls were posted to both loader discords with discussions. 

Now a new discussion will be held on if any other kinds of food categories should be made as well such as fruits or berries. I do want to aim to keep it general and not hyper specific so will be tricky to pin down.

TelepathicGrunt commented:
Summary of new additions. Asked several food modders and people on opinions on new food tags and collected the results. These are several food tags that could prove useful for a good range of mods and would help encourage people to tag their foods properly. Some request for ropes and chains tags came in as quite a few mods are trying to do compat with those but found not very many mods are tagging their ropes and chains. We will see if this default tag will help there.

Also note, a decision was made that if a tag name is a noun that can be plural with or without s/es, we will go with the s/es form. So `c:foods/cooked_fishes` instead of `c:food/cooked_fish`. Let me know if I miss any other previous tags that need their names updated to match this same form where s/es means a group of many different types. Decision for this was due to Minecraft having a `minecraft:fishes` tag and I wanted our raw_ cooked_ fishes tags to match that which then forced the rest of the new tags to also match that convention.

TelepathicGrunt commented:
Note to self: pull in 1.20.5 neoforge branch code and fix merge configs. And patch crafter recipe to read c:player_workstations/crafting_tables tag

Raycoms commented:
I'd like a melee_weapons and ranged_weapons tag.

Ketok4321 commented:
I was randomly looking through changed files of this and fabric's PRs and i think i found an inconsistency: forge has a `c:is_coniferous` biome tag, while fabric has `c:is_tree/coniferous`

TelepathicGrunt commented:
@Ketok4321  Corrected now. This is for NeoForge btw. Not Forge.

KnightMiner commented:
With the number of tags that are being created/suggested that duplicate functionality of tool actions/should also be a tool action, I think it would be worth having a convention for tags that are also tool actions. Something like `c:actions/<tool_action_name>`. This will make it easier to locate a tag with non-conditional items to display in UIs for behavior tied to a tool action, and make it clear when a tag has a tool action alternative that users should use instead when reasonable (as I really don't feel like chasing down every mod that ties their logic to tags instead of tool actions when both exist just so my mod that uses that system works)

sisby-folk commented:
Can goal 4 be corrected to clarify that the translation key format is now `tag.type.namespace.path.by.slashes` e.g. `tag.item.c.ingots.copper` as shown in the current set of changes (and later in the description)?

This is important to note because as you mention later on, having the key be purely derived from the identifier as a convention is what powers tag translations in mods like [EMI](https://github.com/emilyploszaj/emi/blob/1.20/xplat/src/main/resources/assets/emi/lang/en_us.json#L469) - as such implicitly enabling functionality like this for more mods in future by setting the convention here.

TelepathicGrunt commented:
@sisby-folk the code is using the exact same convention from EMI. I asked EMI for their convention so that's what is being used. If you support EMI, you already support this

sisby-folk commented:
In your description under goal 4 there was a typo,  thanks for fixing that - the extra information was just in support of the choice :+1:

Spinoscythe commented:
@TelepathicGrunt since the 1.20.5-pre1 release, we might have to consider adding a `tools` tag as well as it is removed and replaced with `breaks_decorated_pots`

TelepathicGrunt commented:
To answer last question, `c:tools` was not removed from this PR and will remain in since there's a `c:tools/` folder of tags

jacobsjo commented on src/main/java/net/neoforged/neoforge/common/data/internal/NeoForgeBiomeTagsProvider.java:
```diff
@@ -24,71 +25,180 @@ public NeoForgeBiomeTagsProvider(PackOutput output, CompletableFuture<HolderLook
-        tag(Tags.Biomes.IS_WATER).addTag(BiomeTags.IS_OCEAN).addTag(BiomeTags.IS_RIVER);
-        tag(Tags.Biomes.IS_MOUNTAIN).addTag(Tags.Biomes.IS_PEAK).addTag(Tags.Biomes.IS_SLOPE);
+        tag(Tags.Biomes.IS_MOUNTAIN).addTag(Tags.Biomes.IS_MOUNTAIN_PEAK).addTag(Tags.Biomes.IS_MOUNTAIN_SLOPE);
+        tag(Tags.Biomes.IS_AQUATIC).addTag(BiomeTags.IS_OCEAN).addTag(BiomeTags.IS_RIVER);
```
This should reference the `c` variant of the is_ocean and `is_river` tags. (Note: the fabric pr does reference the `c` variants)

jacobsjo commented on src/main/java/net/neoforged/neoforge/common/data/internal/NeoForgeBlockTagsProvider.java:
```diff
@@ -33,40 +34,58 @@ public void addTags(HolderLookup.Provider p_256380_) {
-        tag(Tags.Blocks.GLASS_TINTED).add(Blocks.TINTED_GLASS);
-        addColored(tag(Tags.Blocks.STAINED_GLASS)::add, Tags.Blocks.GLASS, "{color}_stained_glass");
-        tag(Tags.Blocks.GLASS_PANES).addTags(Tags.Blocks.GLASS_PANES_COLORLESS, Tags.Blocks.STAINED_GLASS_PANES);
+        tag(Tags.Blocks.GLASS_BLOCKS).addTags(Tags.Blocks.GLASS_BLOCKS_COLORLESS, Tags.Blocks.GLASS_BLOCKS_CHEAP, Tags.Blocks.GLASS_BLOCKS_TINTED).add(Blocks.WHITE_STAINED_GLASS, Blocks.ORANGE_STAINED_GLASS, Blocks.MAGENTA_STAINED_GLASS, Blocks.LIGHT_BLUE_STAINED_GLASS, Blocks.YELLOW_STAINED_GLASS, Blocks.LIME_STAINED_GLASS, Blocks.PINK_STAINED_GLASS, Blocks.GRAY_STAINED_GLASS, Blocks.LIGHT_GRAY_ ...
```
Nit: This contains `GREEN_STAINED_GLASS` twice.

jacobsjo commented on src/main/java/net/neoforged/neoforge/common/data/internal/NeoForgeBlockTagsProvider.java:
```diff
@@ -33,40 +34,58 @@ public void addTags(HolderLookup.Provider p_256380_) {
+        tag(Tags.Blocks.GLASS_BLOCKS_COLORLESS).add(Blocks.GLASS);
+        tag(Tags.Blocks.GLASS_BLOCKS_CHEAP).add(Blocks.GLASS, Blocks.WHITE_STAINED_GLASS, Blocks.ORANGE_STAINED_GLASS, Blocks.MAGENTA_STAINED_GLASS, Blocks.LIGHT_BLUE_STAINED_GLASS, Blocks.YELLOW_STAINED_GLASS, Blocks.LIME_STAINED_GLASS, Blocks.PINK_STAINED_GLASS, Blocks.GRAY_STAINED_GLASS, Blocks.LIGHT_GRAY_STAINED_GLASS, Blocks.CYAN_STAINED_GLASS, Blocks.PURPLE_STAINED_GLASS, Blocks.BLUE_STAINE ...
+        tag(Tags.Blocks.GLASS_BLOCKS_TINTED).add(Blocks.TINTED_GLASS);
+        tag(Tags.Blocks.GLASS_PANES).addTags(Tags.Blocks.GLASS_PANES_COLORLESS).add(Blocks.WHITE_STAINED_GLASS_PANE, Blocks.ORANGE_STAINED_GLASS_PANE, Blocks.MAGENTA_STAINED_GLASS_PANE, Blocks.LIGHT_BLUE_STAINED_GLASS_PANE, Blocks.YELLOW_STAINED_GLASS_PANE, Blocks.LIME_STAINED_GLASS_PANE, Blocks.PINK_STAINED_GLASS_PANE, Blocks.GRAY_STAINED_GLASS_PANE, Blocks.LIGHT_GRAY_STAINED_GLASS_PANE, Blocks. ...
```
Nit: This contains `GREEN_STAINED_GLASS_PANE` twice.

jacobsjo commented on src/generated/resources/data/c/tags/worldgen/biome/is_windswept.json:
Nit: Not sure how this happens, but this tag contains every element twice.

TelepathicGrunt committed:
Unify tag namespace and conventions between Neoforge and Fabric for 1.20.5 MC

https://github.com/neoforged/NeoForge/pull/135
https://github.com/FabricMC/fabric/pull/3310

TelepathicGrunt committed:
Adjusted untranslated dev log message

TelepathicGrunt committed:
adjust enum name

TelepathicGrunt committed:
renamed enum

TelepathicGrunt committed:
Merge branch 'port/1.20.5' into Common-Tags-Refactor

TelepathicGrunt committed:
pack version updated

TelepathicGrunt committed:
Fixed some tag translation detection issues

TelepathicGrunt committed:
add missing tag translations

TelepathicGrunt committed:
removed unused import

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/common/NeoForgeMod.java:
```diff
@@ -601,6 +605,8 @@ public NeoForgeMod(IEventBus modEventBus, Dist dist, ModContainer container) {
         if (dist.isClient()) ClientCommandHandler.init();
         DualStackUtils.initialise();
+        TagConventionLogWarning.init();
+        if (FMLEnvironment.dist == Dist.CLIENT) TagConventionLogWarningClient.init();
```
The dist is a ctor parameter, and you can probably merge this with the `if` above.

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/common/TagConventionLogWarning.java:
```diff
@@ -0,0 +1,521 @@
+                    if (!legacyTags.isEmpty()) {
+                        StringBuilder stringBuilder = new StringBuilder();
+                        stringBuilder.append("""
+                                \n   Dev warning - Legacy Tags detected. Please migrate your 'forge' namespace tags to 'c' namespace! See net.neoforged.neoforge.common.Tags.java for all tags.
```
```suggestion
                                \n   Dev warning - Legacy Tags detected. Please migrate your 'forge' namespace tags to the 'c' namespace! See net.neoforged.neoforge.common.Tags.java for all tags.
```

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/common/TagConventionLogWarning.java:
```diff
@@ -0,0 +1,521 @@
+                                \n   Dev warning - Legacy Tags detected. Please migrate your 'forge' namespace tags to 'c' namespace! See net.neoforged.neoforge.common.Tags.java for all tags.
+                                   NOTE: Many tags have been moved around or renamed. Some new ones were added so please review the new tags.
+                                    And make sure you follow tag conventions for new tags! The convention is `c` with nouns generally being plural and adjectives being singular.
+                                   You can disable this message in Neoforge's common config by setting logLegacyTagWarnings to "SILENCED" or see individual tags with "DEV_VERBOSE".
```
```suggestion
                                   You can disable this message in NeoForge's common config by setting `logLegacyTagWarnings` to "SILENCED" or see individual tags with "DEV_VERBOSE".
```

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/common/TagConventionLogWarning.java:
```diff
@@ -0,0 +1,521 @@
+                                    TagKey<?> replacementTagkey = LEGACY_FORGE_TAGS.get(tagKey);
+                                    stringBuilder.append("\n     ").append(tagKey).append("  ->  ").append(replacementTagkey);
+                                } else {
+                                    stringBuilder.append("\n     ").append(tagKey).append("  ->  ").append("See similar `c` tags in Neoforge's Tags class");
```
```suggestion
                                    stringBuilder.append("\n     ").append(tagKey).append("  ->  ").append("See similar `c` tags in NeoForge's Tags class");
```

Matyrobbrt reviewed (changes requested):
You have more places where you use `Neoforge` with lowercase f. Please fix those too.

(Leaving review comments is painfully laggy, so I didn't do it for each one.)

TelepathicGrunt committed:
wording and casing fixes
